src/health.cairo

```diff
@@ -0,0 +1,99 @@
+
+    #[abi(embed_v0)]
+    impl HealthCheckImpl of IHealthCheck<ContractState> {
+        fn check_if_healthy(self: @ContractState, gov_address: ContractAddress) -> bool {
+            let proposals_dispatcher = IProposalsDispatcher { contract_address: gov_address };
+            let upgrades_dispatcher = IUpgradesDispatcher { contract_address: gov_address };
+
+            // Check if there are no proposals
+            let current_prop_id = proposals_dispatcher.get_latest_proposal_id();
+            if current_prop_id == 0 {
+                return true;
+            }
+            // Retrieve current proposal details
+            let current_prop_details = proposals_dispatcher.get_proposal_details(current_prop_id);
+
+            // Check if the latest upgrade type matches the proposal's upgrade type
+            let (_, last_upgrade_type) = upgrades_dispatcher.get_latest_upgrade();
+
+            // Ensure that the type of the new proposal matches the required contract type
+            if !self
+                .assert_correct_contract_type(
+                    current_prop_details.to_upgrade, last_upgrade_type.into()
+                ) {
+                return false;
+            }
+
+            // Check the governance state
+            let gov_token_addr = IGovernanceDispatcher { contract_address: gov_address }
+                .get_governance_token_address();
+            let total_eligible_votes_u256: u256 = IERC20Dispatcher {
+                contract_address: gov_token_addr
+            }
+                .total_supply();
+            if total_eligible_votes_u256.high != 0 {
+                return false;
+            }
+            let total_eligible_votes: u128 = total_eligible_votes_u256.low;
+            if total_eligible_votes == 0 {
+                return false;
+            }
+
+            true
+        }
+
+        fn assert_correct_contract_type(
+            self: @ContractState, proposed_contract_type: felt252, previous_contract_type: u64
+        ) -> bool {
+            // Check if the proposed contract type is compatible with the previous contract type
+            if proposed_contract_type == 1 && previous_contract_type == 1 {
+                // Generic contract upgrade is allowed
+                true
+            } else if proposed_contract_type == 3 && previous_contract_type == 3 {
+                // Airdrop upgrade is allowed
+                true
+            } else if proposed_contract_type == 5 && previous_contract_type == 5 {
+                // Custom proposal is allowed
+                true
+            } else if proposed_contract_type == 6 && previous_contract_type == 6 {
+                // Arbitrary proposal is allowed
+                true
+            } else if (proposed_contract_type == 0 && previous_contract_type == 0)
+                || (proposed_contract_type == 2 && previous_contract_type == 2) {
+                false
+            } else {
+                false
+            }
+        }
```

Comment:
I would expect a function with assert in its name to panic if it is not the correct contract type. Either name is is_correct_contract_type and return bool or keep it like this but make it panic if it's not the correct type.

Reply:
I hope this isnt a stupid question-

I know I have to remove the get_latest_prop_id but I'm now confused. Is the way I'm approaching the health check so far the way you were expecting? Or is it completely wrong since I only need to run this for generic proposals. And in this case, what are the checks to ensure that a generic proposal is compatible with the governance state? (would it be things like payload, to_upgrade, or simply that it has a valid quorum, vote time etc.)

Im now confused on what the healthcheck is...